mm2src/coins/z_coin/z_tx_history.rs

```diff
@@ -0,0 +1,222 @@
+);
+
+cfg_native!(
+    use crate::z_coin::BLOCKS_TABLE;
+    use db_common::sqlite::sql_builder::{name, SqlBuilder, SqlName};
+    use db_common::sqlite::rusqlite::Error as SqliteError;
+    use db_common::sqlite::rusqlite::Row;
+    use db_common::sqlite::offset_by_id;
+    use common::async_blocking;
+);
+
+#[cfg(not(target_arch = "wasm32"))]
+const TRANSACTIONS_TABLE: &str = "transactions";
+
+pub(crate) struct ZCoinTxHistoryItem {
+    pub(crate) tx_hash: Vec<u8>,
+    pub(crate) internal_id: i64,
+    pub(crate) height: i64,
+    pub(crate) timestamp: i64,
+    pub(crate) received_amount: i64,
+    pub(crate) spent_amount: i64,
+}
+
+pub(crate) struct ZTxHistoryRes {
+    pub(crate) total_tx_count: u32,
+    pub(crate) transactions: Vec<ZCoinTxHistoryItem>,
+    pub(crate) skipped: usize,
+}
+
+/// Fetch transaction history from the database.
+#[cfg(target_arch = "wasm32")]
+pub(crate) async fn fetch_tx_history_from_db(
+    z: &ZCoin,
+    limit: usize,
+    paging_options: PagingOptionsEnum<i64>,
+) -> Result<ZTxHistoryRes, MmError<ZTxHistoryError>> {
+    let wallet_db = z.z_fields.light_wallet_db.clone();
+    let wallet_db = wallet_db.db.lock_db().await.unwrap();
+    let db_transaction = wallet_db.get_inner().transaction().await?;
+    let tx_table = db_transaction.table::<WalletDbTransactionsTable>().await?;
+    let total_tx_count = tx_table.count_all().await? as u32;
+    let offset = match paging_options {
+        PagingOptionsEnum::PageNumber(page_number) => ((page_number.get() - 1) * limit) as i64,
+        PagingOptionsEnum::FromId(tx_id) => {
+            if tx_id > total_tx_count as i64 {
+                return MmError::err(ZTxHistoryError::FromIdDoesNotExist(tx_id));
+            }
+            (total_tx_count as i64 - tx_id) + 1
+        },
+    };
+
+    // Fetch transactions
+    let txs = tx_table
+        .cursor_builder()
+        .only("ticker", z.ticker())?
+        .offset(offset as u32)
+        .limit(limit)
+        .reverse()
+        .open_cursor("ticker")
+        .await?
+        .collect()
+        .await?;
+
+    // Fetch received notes
+    let rn_table = db_transaction.table::<WalletDbReceivedNotesTable>().await?;
+    let received_notes = rn_table
+        .cursor_builder()
+        .only("ticker", z.ticker())?
+        .open_cursor("ticker")
+        .await?
+        .collect()
+        .await?;
+
+    // Fetch blocks
+    let blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+    let blocks = blocks_table
+        .cursor_builder()
+        .only("ticker", z.ticker())?
+        .open_cursor("ticker")
+        .await?
+        .collect()
+        .await?;
+
+    // Process transactions and construct tx_details
+    let mut tx_details = vec![];
+    for (tx_id, tx) in txs {
+        if let Some((_, WalletDbBlocksTable { height, time, .. })) = blocks
+            .iter()
+            .find(|(_, block)| tx.block.map(|b| b == block.height).unwrap_or_default())
```

Comment:
q: idb doesn't have a join construct like sql?
will such a join ever be expensive?

Reply:
unfortunately indexeddb doesn't offer such abstraction most things are down manually with indexeddb 😄 but indexes they're fast